### DISCUSSION TITLE
TypeScript `typeof number` returns 'object' instead of number

### DISCUSSION THREAD

**Discussion by aimardcr** [vfsfitvnm/frida-il2cpp-bridge#258] [Q&A]
So I notice this really weird behavior, when I'm trying to compare 2 values from the script, if both types are number then the condition would always return false.

This is my code:
```
const SystemData = mainAssembly.class('SystemData'); {
    const m_uiID1 = SystemData.field<number>('m_uiID').value;
    const m_uiID2 = SystemData.field<number>('m_uiID').value;
    console.log(m_uiID1);
    console.log(m_uiID2);
    console.log(typeof m_uiID1);
    console.log(typeof m_uiID2);
    console.log(m_uiID1 === m_uiID2);
}
```

And this is the log output:
```
855084005
855084005
object
object
false
```

However when I'm using `Number(a) == Number(b)` it returns true. Is there something with my code or it's just how TypeScript behave?

**Comment by berkedel** [comment]
What's the output of `console.log(m_uiID1 == m_uiID2)` (loose equality)?

**Reply by aimardcr** [reply]
Still false.

**Comment by vfsfitvnm** [comment]
What's the type of `m_uiID`?

**Reply by aimardcr** [reply]
It's UInt64 for the type

**Reply by vfsfitvnm** [reply]
`System.UInt64` is mapped to [Frida's `UInt64`](https://frida.re/docs/javascript-api/#uint64), so they aren't JS `number`s, but objects

**Reply by aimardcr** [reply]
> `System.UInt64` is mapped to [Frida's `UInt64`](https://frida.re/docs/javascript-api/#uint64), so they aren't JS `number`s, but objects

Ah yeah it does make sense, thanks. I cast it using `Number` constructor and it works fine now, thankyou!

**Reply by vfsfitvnm** [reply]
Beware of doing that. C# `System.UInt64.MaxValue` is higher than JS `Number.MAX_SAFE_INTEGER`. You should use Frida's `UInt64.compare` method instead

**Reply by aimardcr** [reply]
So I make both type `UInt64` instead, the `==` should works fine right?

**Reply by vfsfitvnm** [reply]
```ts
const m_uiID1 = SystemData.field('m_uiID').value;
```
already returns a `UInt64`:
```ts
const m_uiID1 = SystemData.field<UInt64>('m_uiID').value;
```
Hence, you must use `m_uiID1.compare(m_uiID2)` if you want to compare their values

**Reply by aimardcr** [reply]
What if I made both:
```
const m_uiID1 = SystemData.field<UInt64>('m_uiID').value;
const m_uiID2 = SystemData.field<UInt64>('m_uiID').value;
console.log(m_uiID1 == m_uiID2);
```
In theory, it should return true right? i'm on my office pc so i can't test yet.

**Reply by vfsfitvnm** [reply]
No, it returns false. That's why `UInt64.compare` exists

**Reply by aimardcr** [reply]
> No, it returns false. That's why `UInt64.compare` exists

I see, so it will compare the object instead of the value, thank you!
